luaui/Widgets/unit_dgun_stall_assist.lua

```diff
@@ -1,59 +1,88 @@
+		name      = "DGun Stall Assist v2",
 		desc      = "Waits cons/facs when trying to dgun and stalling",
-		author    = "Niobium",
-		date      = "2 April 2010",
+		author    = "zombean",
+		date      = "2 April 2024",
 		license   = "GNU GPL, v2 or later",
 		layer     = 0,
-		enabled   = true  --  loaded by default?
+		enabled   = true
 	}
 end
 
-----------------------------------------------------------------
--- Config
-----------------------------------------------------------------
-local targetEnergy = 600
-local watchForTime = 5
+local millis_start_timer = Spring.GetTimer()
+local function millis()
+  return math.ceil(Spring.DiffTimers(Spring.GetTimer(), millis_start_timer) * 1000)
+end
 
-----------------------------------------------------------------
--- Globals
-----------------------------------------------------------------
-local watchTime = 0
-local waitedUnits = nil -- nil / waitedUnits[1..n] = uID
-local shouldWait = {} -- shouldWait[uDefID] = true / nil
-local isFactory = {} -- isFactory[uDefID] = true / nil
+local target_energy = 1000
+local dgun_cost = 500
+local time_in_ms_befor_unwaiting = 250 --how many fast we can unwait units after waiting them
+local gameStarted = false
+local DGUN_OUT = false
+local UNITS_WAITING = false
+local waitableUnits = {}
+local waitingUnits = {}
+
+local function getWaitAbleUnits()
+  local myWaitableUnits = {}
+  local myUnits = Spring.GetTeamUnits(Spring.GetMyTeamID())
+  for i=1, #myUnits do
+    local unitCmd = Spring.GetUnitCurrentCommand(myUnits[i])
+    local isEnergyConsumingCommand = not (unitCmd == CMD.WAIT or unitCmd == CMD.MOVE or unitCmd == CMD.RECLAIM)
+    if isEnergyConsumingCommand then
+      local unitDefId = Spring.GetUnitDefID(myUnits[i])
+      if waitableUnits[unitDefId] then myWaitableUnits[#myWaitableUnits + 1] = myUnits[i] end
+    end
+  end
+  return myWaitableUnits
+end
 
-local gameStarted
+local function doWeHaveLowEnergy()
+  local currentEnergy, energyStorage = Spring.GetTeamResources(Spring.GetMyTeamID(), "energy")
+  if energyStorage < dgun_cost then return false end
+  if currentEnergy < target_energy then return true end
+  return false
+end
+
+local next_call_when = 0
+local function wait_units()
+  local time_now = millis()
+  if next_call_when > time_now then return end
+  next_call_when = time_now + time_in_ms_befor_unwaiting
+  UNITS_WAITING = true
+  for _, unitID in pairs(getWaitAbleUnits()) do
+    waitingUnits[#waitingUnits +1] = unitID
+  end
+  Spring.GiveOrderToUnitArray(waitingUnits, CMD.WAIT, {}, 0)
+end
+
+local function unwait_units()
+  Spring.GiveOrderToUnitArray(waitingUnits, CMD.WAIT, {}, 0)
+  UNITS_WAITING = false
+  waitingUnits = {}
+end
 
-local stallIds = {UnitDefNames['armcom'].id, UnitDefNames['corcom'].id}
+local function buildWaitableUnitsTable()
+  for uDefID, uDef in pairs(UnitDefs) do
+		if (uDef.buildSpeed > 0) and uDef.canAssist and (not uDef.canManualFire) then
+			waitableUnits[uDefID] = true
+		end
+	end
+end
```

Comment:
Again more of a style thing, but it's generally convention to put this caching process outside of a function near the top of the widget. It will run when the widget loads. Also this code will not catch factories or rezbots
```lua
for uDefID, uDef in pairs(UnitDefs) do
	if (uDef.buildSpeed > 0) and (uDef.canAssist or uDef.canResurrect or uDef.isFactory) and (not uDef.canManualFire) then
		waitableUnits[uDefID] = true
	end
end
```